### ISSUE TITLE
Add async_publish_post event

### ISSUE THREAD

**Issue by mjangda** [Automattic/vip-go-mu-plugins#469]
Similar to what we have on WP.com (see `queue_publish_post`).

Whenever a post is updated we could queue up a cron event that can be used to perform various non-essential tasks.

Some core examples include `update_posts_count` and `wpmu_update_blogs_date`. Some Go-specific ones could be cache purges and `lastpostmod` bumps. VIPs could hook into this as well for their own heavy events/actions.

**Comment by ethitter** [comment]
For sites running Cron Control, we can treat this as a "internal event," which will give it priority and ensure reliable execution. Sites running WP-Cron Control should be fine with this as well. We do have a few sites that rely on traditional WordPress cron execution, which we'll want to exempt lest these async jobs never run.

**Comment by mjangda** [comment]
Some more misc thoughts:

- WP.com only queues jobs for `publish_post`. Do we want to handle more status transitions (e.g. publish => trash)? 
- Should we queue one event where others hook in and we execute them all at the same time? (could result in cron events that take a long time to complete) Or should we offer a handler than allows queueing multiple events per task? (could result in lots of cron events per post update)

**Comment by ethitter** [comment]
> WP.com only queues jobs for `publish_post`. Do we want to handle more status transitions (e.g. publish => trash)?

Seems worth queueing the job whenever either the new or old status is `publish` (skipping when they're both publish), since post counts and other things would change in that situation. We should be sure to cover CPTs, at least public CPTs, as `post` isn't guaranteed to be the primary CPT for every site.

> Should we queue one event where others hook in and we execute them all at the same time? Or should we offer a handler than allows queueing multiple events per task?

I'd rather queue discrete events. Too many long-running tasks in a single event are likely to run into the time limits imposed by our various cron runners, and separate entries will simplify debugging.

**Comment by ethitter** [comment]
I changed my mind a bit while revising #471. 😄 

I opted for one job, because it's more natural for users to hook to an action. It also means that a lot of code can be moved to the hook with the addition of `async_` to the action. 

Right now, it applies to all post types. Should it?

Should transitioning from published to unpublished trigger a different action? Right now, it still fires `async_publish_post`.

Should we provide a filter to disable it for post types, statuses, etc?

Might something be missed, since we don't requeue if there's already a scheduled event?
